# Part 1. Equation of a Slime

How many late days are you using for this assignment? None

In [172]:
# Imports section
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold


## 1. Loading the dataset

In [173]:
# Using pandas load the dataset
science = pd.read_csv("science_data_large.csv")
ckd = pd.read_csv("ckd_feature_subset.csv")


# Output the first 15 rows of the data
print(science.head(15))
print(ckd.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(science.info())
print(ckd.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
         age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0   0.688312  0.333333  0.000000           1           0  0.000000           1
1   0.545455  0.333333  0.128319           1           0  0.305085           1
2   0.714286  0.500000  0.238938           1           0  0.186441         

## 2. Splitting the dataset

In [174]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = science.drop(columns=['Size nm^3'])
y = science['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# For grading consistency use random_state=42 

## 3. Perform a Linear Regression

In [175]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = np.array([X_test.iloc[0]]).reshape(1, -1)
prediction = model.predict(sample_data)
print("Sample Prediction:", prediction)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print("Linear Regression Score:", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print("Coefficients:", model.coef_)
print("Intercept:", model.intercept_)

Sample Prediction: [235911.1927226]
Linear Regression Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.4795834075


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)


In [176]:
print(f"Equation of Slime Growth: f(x) = {model.intercept_:.5f} + " + " + ".join([f"{coeff:.5f}*x{i+1}" for i, coeff in enumerate(model.coef_)]))

Equation of Slime Growth: f(x) = -409391.47958 + 866.14641*x1 + 1032.69507*x2


Report on score and explain meaning:

The score represents the coefficient of determination (or R^2 score) which measures how well the model explains the variance in the data.
A high R^2 value and the closer it is to 1 indicates that the model explains most of the variability in the response variable. Since our score is 0.855, it is rather close to 1 showing that the model explains most of the variability in the slime size data.

## 4. Use Cross Validation

In [177]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

cv_scores = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(model, X, y, cv=cv_scores, scoring='r2')

# Report on their finding and their significance
print("cross validation scores:", scores)
print("mean CV score:", np.mean(scores))
print("standard deviation", np.std(scores))

cross validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
mean CV score: 0.8597294202684646
standard deviation 0.01838773713930639


Write findings here:

Cross validation helps test the stability of the model across different subsets of the data. The average of the cross validation score is 0.8597, which is pretty close to our R^2 score. This means that the model is generalizing well to new data and is consistent across different folds. The standard deviation is rather small which also shows our model is consistent.


## 5. Using Polynomial Regression

In [178]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
polynomial_model = make_pipeline(PolynomialFeatures(2), LinearRegression())
cv_scores_polynomial = cross_val_score(polynomial_model, X, y, cv=5)
print("polynomial regression mean CV score:", np.mean(cv_scores_polynomial))

# Report on the metrics and output the resultant equation as you did in Part 3.

polynomial regression mean CV score: 1.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

In [ ]:
print(f"polynomial equation of : f(x) = {model.intercept_:.5f} + " + " + ".join([f"{coeff:.5f}*x{i+1}" for i, coeff in enumerate(model.coef_)]))

Equation of Slime Growth: f(x) = -409391.47958 + 866.14641*x1 + 1032.69507*x2


Report on the score and interpret:

The mean cross validation score for polynomial regression is 1, which is higher than the linear model. The model’s predictions were all correct for the interpretation since it’s 1 for all the folds.


# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [161]:
# Load the dataset. Then train and evaluate the classification models.


X = ckd.drop(columns=['Target_ckd'])  
y = ckd['Target_ckd']

X_train_ckd, X_test_ckd, y_train_ckd, y_test_ckd = train_test_split(X, y, test_size=0.1, random_state=42)

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

models = {
    "Logistic Regression": LogisticRegression(random_state=42, max_iter=1000),
    "SVM": SVC(kernel='rbf', random_state=42),
    "k-NN": KNeighborsClassifier(n_neighbors=5),
    "Neural Network": MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=42)
}

results = {}
for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')
    results[name] = {
        "mean accuracy": np.mean(scores),
        "standard deviation": np.std(scores)
    }


results_df = pd.DataFrame(results).T
print(results_df)


                     mean accuracy  standard deviation
Logistic Regression       0.856129            0.054236
SVM                       0.928172            0.059993
k-NN                      0.928172            0.031893
Neural Network            0.941290            0.024040


## Results and Conclusion for Classification Experiments

In [162]:
# Experiments with Neural Network.
nn_configs = {
    "Small NN": MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=42),
    "Medium NN": MLPClassifier(hidden_layer_sizes=(20, 20), max_iter=1000, random_state=42),
    "Large NN": MLPClassifier(hidden_layer_sizes=(50, 50, 50), max_iter=1000, random_state=42)
}


res = {}

for name, model in nn_configs.items():
    scores = cross_val_score(model, X_scaled, y, cv=5)
    res[name] = {
        "mean accuracy": np.mean(scores),
        "standard deviation": np.std(scores)
    }


res_df = pd.DataFrame(res).T
print(res_df)

/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


           mean accuracy  standard deviation
Small NN        0.967312            0.021088
Medium NN       0.973763            0.024835
Large NN        0.960860            0.024610


## Results and Conclusion for Neural Network Experiments

According to the experiment results, Neural Networks is the most accurate model with a mean accuracy of 0.941 which was the highest of all 4. It performed the best with a medium sized neural network with a mean accuracy of 0.974. The worst performing model was the Logistic Regression with a mean accuracy of 0.8561.